### PR TITLE
[FIX] mrp_subcontracting: fix flow with tracked component

### PR DESCRIPTION
**Pull request by ryv-odoo**
After the odoo/odoo#52949 , the registering of tracked component,
in case of subcontracting, was broken. Fix and try to mimic
the same flow than before. The produce wizard has been replaced
by a clean MO form to register component and use backorder mechanisme
of MO to manage tracked finished product + tracked component.

task-2278147

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp_subcontracting/models/mrp_production.py [hunk 1]
```diff
@@ -1,20 +1,64 @@
...
+            line_by_product = defaultdict(lambda: self.env['stock.move.line'])
+            for line in production.move_line_raw_ids:
+                line_by_product[line.product_id] |= line
+            for move in production.move_raw_ids:
+                move.move_line_ids = line_by_product.pop(move.product_id, self.env['stock.move.line'])
+            for product_id, lines in line_by_product.items():
+                qty = sum(line.product_uom_id._compute_quantity(line.qty_done, product_id.uom_id) for line in lines)
+                move = production._get_move_raw_values(product_id, qty, product_id.uom_id)
+                move['additional'] = True
+                production.move_raw_ids = [(0, 0, move)]
```
Comment (amoyaux) [inline addons/mrp_subcontracting/models/mrp_production.py:34]: move['move_line_ids'] = [(4, ml.id) for ml in lines]
and remove
production.move_raw_ids.filtered(lambda m: m.product_id == product_id)[:1].move_line_ids = lines
?

File: addons/mrp_subcontracting/models/mrp_production.py [hunk 2]
```diff
@@ -1,20 +1,64 @@
...
+                production.move_raw_ids.filtered(lambda m: m.product_id == product_id)[:1].move_line_ids = lines
+
+    def subcontracting_record_component(self):
+        self.ensure_one()
+        assert self.env.context.get('subcontract_move_id')
+        if float_is_zero(self.qty_producing, precision_rounding=self.product_uom_id.rounding):
+            return {'type': 'ir.actions.act_window_close'}
+        for sml in self.move_raw_ids.move_line_ids:
+            if sml.tracking != 'none' and not sml.lot_id:
+                raise UserError(_('You must enter a serial number for each line of %s') % sml.product_id.name)
```
Comment (amoyaux) [inline addons/mrp_subcontracting/models/mrp_production.py:44]: _('You must enter a serial number for each line of %s', sml.product_id.name)

File: addons/mrp_subcontracting/models/stock_move.py [hunk 3]
```diff
@@ -54,8 +54,7 @@ def write(self, values):
         if 'product_uom_qty' in values:
             if self.env.context.get('cancel_backorder') is False:
                 return super(StockMove, self).write(values)
-            self.filtered(lambda m: m.is_subcontract and
-            m.state not in ['draft', 'cancel', 'done'])._update_subcontract_order_qty(values['product_uom_qty'])
+            self.filtered(lambda m: m.is_subcontract and m.state not in ['draft', 'cancel', 'done'])._update_subcontract_order_qty(values['product_uom_qty'])
```
Comment (amoyaux) [inline addons/mrp_subcontracting/models/stock_move.py:57]: Should only update one of the production

File: addons/mrp_subcontracting/models/mrp_production.py (hunk 1 above)
Comment (ryv-odoo) [reply addons/mrp_subcontracting/models/mrp_production.py:34]: Obviously, It doesn't work. I don't know the real reason behind that :/

File: addons/mrp_subcontracting/models/mrp_production.py (hunk 2 above)
Comment (ryv-odoo) [reply addons/mrp_subcontracting/models/mrp_production.py:44]: Only in master

File: addons/mrp_subcontracting/tests/test_subcontracting.py [hunk 4]
```diff
@@ -686,7 +603,85 @@ def test_flow_tracked_1(self):
...
+                'product_id': self.finished_product.id,
+                'company_id': self.env.company.id,
+            }))
+            serials_comp1.append(self.env['stock.production.lot'].create({
+                'name': 'serials_comp1_%s' % i,
+                'product_id': self.comp1_sn.id,
+                'company_id': self.env.company.id,
+            }))
+
+        # TODO : Waiting fix for (2, 0) in test form fix
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/tests/test_subcontracting.py:657]: was the fix merged ?

Comment (ryv-odoo) [reply addons/mrp_subcontracting/tests/test_subcontracting.py:657]: In the testing environment, no :/

Comment (ryv-odoo) [reply addons/mrp_subcontracting/tests/test_subcontracting.py:657]: My bad, it is merged in https://github.com/odoo/odoo/pull/54642/commits/fae2ecd580ff2a9c67de25ea04378858a4633a8f
I will uncomment this part

**Comment by sle-odoo** [comment]
robodoo r+

**Comment by sle-odoo** [comment]
robodoo r+

**Comment by sle-odoo** [comment]
robodoo r+

**Pull request closed without merging**